wave_matching/src/icp.cpp

```diff
@@ -84,12 +130,152 @@ void ICPMatcher::estimateInfo() {
+                               this->params.max_corr)  // if the distance to
+                                                       // point is less than max
+                                                       // correspondence
+                                                       // distance, use it to
+                                                       // calculate
+        {
+            Eigen::Vector3f source_pt = qpt.getVector3fMap();
+            Eigen::Vector3f target_pt =
+              (*targetc)->points[nn_idx[0]].getVector3fMap();
+
+            // Compute the point pair average and difference and store for later
+            // use
+            corrs_aver[numCorr] = 0.5 * (source_pt + target_pt);
+            corrs_diff[numCorr] = source_pt - target_pt;
+            numCorr++;
+        } else {
+            continue;
+        }
+    }
+    corrs_aver.resize(numCorr);
+    corrs_diff.resize(numCorr);
+
+    // now compute the M matrix
+    wave::Mat6 MM = wave::Mat6::Zero();
+    wave::Vec6 MZ = wave::Vec6::Zero();
+    for (int ci = 0; ci != numCorr; ++ci)  // ci = correspondence iterator
+    {
+        // Fast computation of summation elements of M'M
+        MM(0, 4) -= corrs_aver[ci](1);
+        MM(0, 5) += corrs_aver[ci](2);
+        MM(1, 3) -= corrs_aver[ci](2);
+        MM(1, 4) += corrs_aver[ci](0);
+        MM(2, 3) += corrs_aver[ci](1);
+        MM(2, 5) -= corrs_aver[ci](0);
+        MM(3, 4) -= corrs_aver[ci](0) * corrs_aver[ci](2);
+        MM(3, 5) -= corrs_aver[ci](0) * corrs_aver[ci](1);
+        MM(4, 5) -= corrs_aver[ci](1) * corrs_aver[ci](2);
+        MM(3, 3) += corrs_aver[ci](1) * corrs_aver[ci](1) +
+                    corrs_aver[ci](2) * corrs_aver[ci](2);
+        MM(4, 4) += corrs_aver[ci](0) * corrs_aver[ci](0) +
+                    corrs_aver[ci](1) * corrs_aver[ci](1);
+        MM(5, 5) += corrs_aver[ci](0) * corrs_aver[ci](0) +
+                    corrs_aver[ci](2) * corrs_aver[ci](2);
+
+        // Fast computation of M'Z
+        MZ(0) += corrs_diff[ci](0);
+        MZ(1) += corrs_diff[ci](1);
+        MZ(2) += corrs_diff[ci](2);
+        MZ(3) += corrs_aver[ci](1) * corrs_diff[ci](2) -
+                 corrs_aver[ci](2) * corrs_diff[ci](1);
+        MZ(4) += corrs_aver[ci](0) * corrs_diff[ci](1) -
+                 corrs_aver[ci](1) * corrs_diff[ci](0);
+        MZ(5) += corrs_aver[ci](2) * corrs_diff[ci](0) -
+                 corrs_aver[ci](0) * corrs_diff[ci](2);
+    }
+    // Remaining elements of M'M
+    MM(0, 0) = MM(1, 1) = MM(2, 2) = static_cast<float>(numCorr);
+    MM(4, 0) = MM(0, 4);
+    MM(5, 0) = MM(0, 5);
+    MM(3, 1) = MM(1, 3);
+    MM(4, 1) = MM(1, 4);
+    MM(3, 2) = MM(2, 3);
+    MM(5, 2) = MM(2, 5);
+    MM(4, 3) = MM(3, 4);
+    MM(5, 3) = MM(3, 5);
+    MM(5, 4) = MM(4, 5);
+
+    // Compute pose difference estimation
+    wave::Vec6 D = static_cast<wave::Vec6>(MM.inverse() * MZ);
+
+    // Compute s^2
+    float ss = 0.0f;
```

Comment:
Sorry for nitpicking ben. Do you mind being a bit more specific in your variable names? M, Z, s, and D are a little bit ambiguous.

Reply:
These names follow the convention in a Lu & Milios paper. I will put a comment saying as much.

Reply:
And where to find paper